### ISSUE TITLE
[miopen-hip/opencl] Build failure boost inline

### ISSUE THREAD

**Issue by bytbox** [rocm-arch/rocm-arch#511]
When I try to build (using the aur PKGBUILD, nothing changed), I get error messages originating from boost headers, centering around the keyword `noinline`. See the build log here: https://gist.github.com/bytbox/019ab8439b83cf3edd7965c2265187a5

The closest thing I can find online is [this](https://github.com/BVLC/caffe/issues/171). That's old enough that I kind of doubt it's related.

No -testing repositories are enabled for me. This is my first attempt to build, so I have no idea if this is a new problem (and also it's likely just me doing something wrong).

Hardware (surely not relevant): Navi 10 [Radeon RX 5600 OEM/5600 XT / 5700/5700 XT]

**Comment by QMechanic** [comment]
Today I tried re-installing miopen-hip again after I updated my kernel to 5.10.0-1-mainline, just in case for some reason something would change from my previous issue (#509), and surprisingly it was different. I got exactly the same issue as you.
Now I thought that was something I did or maybe due to my GPU not being supported (RX 6800) because I highly doubt it would be the kernel (but I also can't find anything which changed in this repository that would change that so I am unsure, possibly a new version of boost I installed without remembering).

Now I am mostly commenting to say that you're not the only person and that I have exactly the same errors and sadly have no solution.

But I am wondering if anyone has gotten miopen-hip to work on any of the GPUs from 5000 series or later (gfx10xx) because from what I have read it seems like support for those has been really bad, especially regarding the machine learning area (for example they are not even on the list for officially supported GPUs on https://github.com/RadeonOpenCompute/ROCm#Hardware-and-Software-Support).

It would be great if someone could verify/clarify if this is an issue of hardware support or something else.

**Comment by bytbox** [comment]
@QMechanic thanks for chiming in -- always nice to know I'm not completely crazy.

My kernel is merely 5.9.14-arch1-1. Looking at your build log, your previous build was on boost 1.74. I'm guessing that your upgrade took boost to 1.75 as well -- can you confirm?

That being said, it's not obvious this is a bug in boost. I note the following in my error:

`
/opt/rocm/hip/include/hip/hcc_detail/host_defines.h:75:46: note: expanded from macro '__noinline__'
#define __noinline__ __attribute__((noinline))
`

The problematic line is within an `#ifdef __HCC__`. Perhaps `__HCC__` is getting set when it shouldn't be. Or, perhaps the wrong compiler is being called.

**Comment by QMechanic** [comment]
I just checked, boost and boost-libs are in fact 1.75.0-1 now. I didn't realize you could see the old version in the build log but that's great then we know something more.

Remember that I have the new GPU (RX 6800, gfx1030, ID: 0x73bf) which is not officially supported (honestly I believe this is the issue, but not exactly sure why it would break).
In case you or someone else wants to compare my logs that I get now with yours for some reason here it is https://gist.github.com/QMechanic/7722aafd4b2c727bed20347e4ecd1e52
If there are any tests I can do to help out then share.

Hopefully everyone will be able to use their GPUs for machine learning after ROCm 4.0 comes out.

**Comment by bytbox** [comment]
Vaguely related: https://svn.boost.org/trac10/ticket/9392. (Not helpful though.)

It looks like there's a conflict over whether `__noinline__` should be an alias for `noinline` or an alias for `__attribute__ ((noinline))`. I think we can all agree that it shouldn't be an alias for `__attribute__ ((__attribute__ (noinline) ))`, though!

I can get that particular file to compile by adding `-D BOOST_NOINLINE="__attribute__((noinline))"` to the hipcc invocation. This is probably not the correct generic fix (although maybe for arch packaging purposes, it's sufficient?). I also can't figure out where to insert this to get cmake to always use that setting. I'll take another stab tomorrow.

(As far as support goes... see this conversation: https://github.com/RadeonOpenCompute/ROCm/issues/887. My card is some navi thing, so as far as I understand, even if I get the stack compiled, I won't be able to use pytorch. This is pretty disappointing. OTOH, they must need civilian testers...)

**Comment by takfate** [comment]
sorry. I have a question: does this package support `AMD CPU`?

**Comment by takfate** [comment]
And I have the following error when compiling.
```
[3647/4636] cd /var/tmp/pamac-build-tak/python-pytorch-rocm/src/pytorch-1.7.1-rocm/build/caffe2/CMakeFiles/torch_hip.dir/__/aten/src/THH && /usr/bin/cmake -E make_directory /var/tmp/pamac-build-tak/python-pytorch-rocm/src/pytorch-1.7.1-rocm/build/caffe2/CMakeFiles/torch_hip.dir/__/aten/src/THH/. && /usr/bin/cmake -D verbose:BOOL=OFF -D build_configuration:STRING=RELEASE -D generated_file:STRING=/var/tmp/pamac-build-tak/python-pytorch-rocm/src/pytorch-1.7.1-rocm/build/caffe2/CMakeFiles/torch_hip.dir/__/aten/src/THH/./torch_hip_generated_THHTensorIndex.hip.o -P /var/tmp/pamac-build-tak/python-pytorch-rocm/src/pytorch-1.7.1-rocm/build/caffe2/CMakeFiles/torch_hip.dir/__/aten/src/THH/torch_hip_generated_THHTensorIndex.hip.o.cmake
ninja: build stopped: subcommand failed.
Building wheel torch-1.7.1
-- Building version 1.7.1
cmake -GNinja -DBUILD_BINARY=ON -DBUILD_CUSTOM_PROTOBUF=OFF -DBUILD_PYTHON=True -DBUILD_TEST=True -DCMAKE_BUILD_TYPE=Release -DCMAKE_INSTALL_PREFIX=/var/tmp/pamac-build-tak/python-pytorch-rocm/src/pytorch-1.7.1-rocm/torch -DCMAKE_PREFIX_PATH=/usr/lib/python3.8/site-packages -DCUDNN_INCLUDE_DIR=/usr/include -DCUDNN_LIBRARY=/usr/lib -DDISABLE_AVX2=1 -DDISABLE_AVX512F=1 -DDISABLE_FMA4=1 -DDISABLE_SSE4=1 -DNUMPY_INCLUDE_DIR=/usr/lib/python3.8/site-packages/numpy/core/include -DPYTHON_EXECUTABLE=/usr/bin/python -DPYTHON_INCLUDE_DIR=/usr/include/python3.8 -DPYTHON_LIBRARY=/usr/lib/libpython3.8.so.1.0 -DTORCH_BUILD_VERSION=1.7.1 -DUSE_CUDA=OFF -DUSE_FFMPEG=ON -DUSE_GFLAGS=ON -DUSE_GLOG=ON -DUSE_MKLDNN=ON -DUSE_NUMPY=True -DUSE_ROCM=ON -DUSE_SYSTEM_NCCL=ON /var/tmp/pamac-build-tak/python-pytorch-rocm/src/pytorch-1.7.1-rocm
cmake --build . --target install --config Release -- -j 16
Traceback (most recent call last):
  File "setup.py", line 760, in <module>
    build_deps()
  File "setup.py", line 310, in build_deps
    build_caffe2(version=version,
  File "/var/tmp/pamac-build-tak/python-pytorch-rocm/src/pytorch-1.7.1-rocm/tools/build_pytorch_libs.py", line 62, in build_caffe2
    cmake.build(my_env)
  File "/var/tmp/pamac-build-tak/python-pytorch-rocm/src/pytorch-1.7.1-rocm/tools/setup_helpers/cmake.py", line 349, in build
    self.run(build_args, my_env)
  File "/var/tmp/pamac-build-tak/python-pytorch-rocm/src/pytorch-1.7.1-rocm/tools/setup_helpers/cmake.py", line 141, in run
    check_call(command, cwd=self.build_dir, env=env)
  File "/usr/lib/python3.8/subprocess.py", line 364, in check_call
    raise CalledProcessError(retcode, cmd)
subprocess.CalledProcessError: Command '['cmake', '--build', '.', '--target', 'install', '--config', 'Release', '--', '-j', '16']' returned non-zero exit status 1.
==> ERROR: A failure occurred in build().
    Aborting...
```

**Comment by acxz** [comment]
@takfate it seems your question and error log are not related to the build error we are discussing in this issue. Can you please open a separate issue for them?

**Comment by petronny** [comment]
This error can be reproduced when building in clean chroot.

**Comment by tpkessler** [comment]
I can reproduce this with MIOpen 4.0.0, see https://github.com/ROCmSoftwarePlatform/MIOpen/issues/658

**Comment by nquest** [comment]
Just rebuilt with downgrade boost to 1.72 as @tpkessler mentioned in https://github.com/ROCmSoftwarePlatform/MIOpen/issues/658 and again same error about rocblas.h not found

**Comment by acxz** [comment]
Let's just worry about building with the boost that arch provides, don't really want to worry about adding support for older boosts. If it doesn't work with both versions, let focus on fixing it for the latest version.

**Comment by JP-Ellis** [comment]
I have been having the same issue and found that the issue can be avoided by enabling the `BOOST_CONTAINER_DISABLE_NOINLINE` flag. This flag was added in boostorg/container@c08ead7f4b4e4aa5674151e180f05fcc0a05e644 but is otherwise entirely undocumented.

Although it removes the main issue, the linking of `libMIOpen.so` fails later due ot the following error:

```
/sbin/ld: CMakeFiles/MIOpen.dir/mdg_expr.cpp.o: in function `boost::spirit::utree boost::spirit::function_impl_invert::eval<bool>(bool const&)':
mdg_expr.cpp:(.text+0x0): multiple definition of `boost::spirit::utree boost::spirit::function_impl_invert::eval<bool>(bool const&)'; CMakeFiles/MIOpen.dir/md_graph.cpp.o:md_graph.cpp:(.text+0x0): first defined here
clang-12: error: linker command failed with exit code 1 (use -v to see invocation)
make[2]: *** [src/CMakeFiles/MIOpen.dir/build.make:2662: lib/libMIOpen.so.1.0] Error 1
make[2]: Leaving directory '/tmp/makepkg/miopen-hip/src/build'
make[1]: *** [CMakeFiles/Makefile2:2636: src/CMakeFiles/MIOpen.dir/all] Error 2
make[1]: Leaving directory '/tmp/makepkg/miopen-hip/src/build'
make: *** [Makefile:182: all] Error 2
make: Leaving directory '/tmp/makepkg/miopen-hip/src/build'
```

I don't know whether this later error is related to the addition of the flag yet.

**Comment by acxz** [comment]
Good find @JP-Ellis 
The error you posted seems similar to the one we saw before boost was updated on Arch, #509 

we can add the patched label once we get a PR solution for this one merged in. Would you like to submit a PR based on your findings?

**Comment by tpkessler** [comment]
This linker error also appears when compiling `miopen-opencl`. This issue was already reported upstream https://github.com/ROCmSoftwarePlatform/MIOpen/issues/408 and is related to a change in boost 1.74 https://github.com/boostorg/spirit/commit/36f8e09cfd112840085472627a6f2b9e11ff5739#diff-316724f0e450faeac4f414af9c8b17bd

**Comment by acxz** [comment]
reopening as the problem still persists with *-opencl.

**Comment by acxz** [comment]
patched with #556 and #559

**Comment by JP-Ellis** [comment]
This can probably be closed as it has been patched on both versions of `miopen`.  #509 can keep track of the remaining error during linking.

**Comment by acxz** [comment]
@JP-Ellis we keep 'patched' issues open until upstream releases the fix. Until we don't have to build with the boost flag or until upstream documentation states that is how they recommend building miopen, the "patch" will be necessary and hence this issue remains open to track it the patched status.

**Comment by tpkessler** [comment]
I've updated the package according to recommendations upstream. It now compiles successfully in a clean chroot. Can you confirm that @bytbox @QMechanic @JP-Ellis ?

**Comment by bytbox** [comment]
It builds fine for me (no clean anything, it just works). Thanks! (And now to figure out if pytorch can be gotten to work...)

Closing.

**Comment by JP-Ellis** [comment]
@tpkessler Thanks for the update!  However I noticed that the PKGBUILD now downloads and installs its own dependencies without using the system ones.  Is this really what we want to do though?  Does it really need to compile it's own version of `sqlite3`, `bzip2` and `boost`?

The reason why this works is not because the upstream issue has been fixed, but because you've reverted back to using version `1.72.0` of Boost (the current one on Arch Linux is `1.75.0`), before the bug was introduced.

Do you intend this to be only a temporary fix?  I'm happy to open a tracking another tracking issue regarding this so that when Boost is next updated, we can start using the latest version of Boost again.

**Comment by nquest** [comment]
@JP-Ellis , upstream issue in ROCm closed, because error in boost since 1.74 so use embedded boost is an option.

**Comment by tpkessler** [comment]
Hey @JP-Ellis, I agree with you that building `boost` from scratch is not optimal but there are two reasons for doing this.

1. MIOpen has pinned the `boost` version to `1.72`:
>  Boost at version 1.72 (other versions are not supported) 

(from https://github.com/ROCmSoftwarePlatform/MIOpen)
The recommend way to install `boost 1.72` upstream is via `cmake`, see [here](https://rocmsoftwareplatform.github.io/MIOpen/doc/html/install.html#installing-minimum-dependencies-in-rocm-environment).

2. The `noinline` issue is caused by a bug in HIP:
> Actually, this is a bug in HIP, as its defining a \_\_noinline\_\_ macro and boost is trying to use the attribute \_\_noinline\_\_ causing the error.

(from https://github.com/ROCmSoftwarePlatform/MIOpen/issues/658#issuecomment-754702931)

Once AMD has fixed the HIP bug we can try to build `miopen` with system `boost` again. But as of now, we stick with `boost 1.72`.

> I'm happy to open a tracking another tracking issue regarding this so that when Boost is next updated, we can start using the latest version of Boost again.

I would appreciate that.
